datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -534,6 +574,655 @@ impl LogicalPlan {
+                table_name,
+                source,
+                projection,
+                projected_schema,
+                filters,
+                fetch,
+            }) => {
+                let filters = filters
+                    .iter()
+                    .map(|f| {
+                        Self::replace_placeholders_with_values(f.clone(), param_values)
+                            .expect("Failed to replace params in table scan filter")
+                    })
+                    .collect();
+                Ok(LogicalPlan::TableScan(TableScan {
+                    table_name: table_name.clone(),
+                    source: Arc::clone(source),
+                    projection: projection.clone(),
+                    projected_schema: Arc::clone(projected_schema),
+                    filters,
+                    fetch: *fetch,
+                }))
+            }
+            LogicalPlan::Values(Values { values, schema }) => {
+                let values = values
+                    .iter()
+                    .map(|row| {
+                        row.iter()
+                            .map(|expr| {
+                                Self::replace_placeholders_with_values(
+                                    expr.clone(),
+                                    param_values,
+                                )
+                                .expect("Failed to replace params in values")
+                            })
+                            .collect()
+                    })
+                    .collect();
+                Ok(LogicalPlan::Values(Values {
+                    values,
+                    schema: Arc::clone(schema),
+                }))
+            }
+            LogicalPlan::EmptyRelation(EmptyRelation {
+                produce_one_row,
+                schema,
+            }) => Ok(LogicalPlan::EmptyRelation(EmptyRelation {
+                produce_one_row: *produce_one_row,
+                schema: Arc::clone(schema),
+            })),
+            LogicalPlan::Explain(_)
+            | LogicalPlan::Analyze(_)
+            | LogicalPlan::CreateMemoryTable(_)
+            | LogicalPlan::CreateView(_)
+            | LogicalPlan::CreateExternalTable(_)
+            | LogicalPlan::CreateCatalogSchema(_)
+            | LogicalPlan::CreateCatalog(_)
+            | LogicalPlan::DropTable(_)
+            | LogicalPlan::SetVariable(_)
+            | LogicalPlan::DropView(_) => Err::<LogicalPlan, DataFusionError>(
+                DataFusionError::NotImplemented(format!(
+                    "This logical plan should not contain parameters/placeholder: {}",
+                    self.display()
+                )),
+            ),
+        }
+    }
+
+    /// Recrusively to walk the expression and convert a placeholder into a literal value
+    fn replace_placeholders_with_values(
+        expr: Expr,
+        param_values: &Vec<ScalarValue>,
+    ) -> Result<Expr, DataFusionError> {
```

Comment:
I think you could use an `ExprRewriter here to avoid having to write your own recursion for each type of expr, making this code shorter and easier to maintain

https://docs.rs/datafusion-expr/15.0.0/datafusion_expr/expr_rewriter/trait.ExprRewriter.html


Something like this perhaps

https://github.com/apache/arrow-datafusion/blob/31bbe6c3eec245127b005d8afc8766e2d06d811b/datafusion/optimizer/src/push_down_filter.rs#L748-L767